powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/ActivationServiceBehavior.java

```diff
@@ -1976,8 +1992,8 @@ public void expireOperations() {
         final ImmutableSet<ActivationStatus> activationStatuses = ImmutableSet.of(ActivationStatus.CREATED, ActivationStatus.PENDING_COMMIT);
         try (final Stream<ActivationRecordEntity> abandonedActivations = activationRepository.findAbandonedActivations(activationStatuses, lookBackTimestamp, currentTimestamp)) {
             abandonedActivations.forEach(activation -> {
-                logger.info("Removing abandoned activation with ID: {}, revoking recovery codes.", activation.getActivationId());
-                deactivatePendingActivation(currentTimestamp, activation, false);
+                logger.info("Removing abandoned activation with ID: {}", activation.getActivationId());
+                removeActivation(activation, null, true);
```

Comment:
removeActivation does not have same timestap guarding mechanism as deactivatePendingActivation. Are we not removing activation before expiration? Or this was redundant, as there is a special query for finding abandoned/expired Activations

Reply:
This is redundant due to way we search for activations but I tend to agree we should keep `deactivatePendingActivation`, since we use this method elsewhere for deactivating activations that were abandoned.

Another thing to consider is to cleanup the code since the methods are quite similar:

```java
private void deactivatePendingActivation(Date timestamp, ActivationRecordEntity activation, boolean isActivationLocked) {
    if ((activation.getActivationStatus().equals(ActivationStatus.CREATED) || activation.getActivationStatus().equals(ActivationStatus.PENDING_COMMIT)) && (timestamp.getTime() > activation.getTimestampActivationExpire().getTime())) {
        logger.info("Deactivating pending activation, activation ID: {}", activation.getActivationId());
        if (!isActivationLocked) {
            // Make sure activation is locked until the end of transaction in case it was not locked yet
            activation = repositoryCatalogue.getActivationRepository().findActivationWithLock(activation.getActivationId());
        }
        activation.setActivationStatus(io.getlime.security.powerauth.app.server.database.model.ActivationStatus.REMOVED);
        revokeRecoveryCodes(activation.getActivationId());
        activationHistoryServiceBehavior.saveActivationAndLogChange(activation);
        callbackUrlBehavior.notifyCallbackListenersOnActivationChange(activation);
    }
}
```

```java
public RemoveActivationResponse removeActivation(@NotNull ActivationRecordEntity activation, String externalUserId, boolean revokeRecoveryCodes) {
    logger.info("Processing activation removal, activation ID: {}", activation.getActivationId());
    activation.setActivationStatus(io.getlime.security.powerauth.app.server.database.model.ActivationStatus.REMOVED);
    // Recovery codes are revoked in case revocation is requested, or always when the activation is in CREATED or PENDING_COMMIT state
    if (revokeRecoveryCodes ||
            (activation.getActivationStatus() == ActivationStatus.CREATED ||
                    activation.getActivationStatus() == ActivationStatus.PENDING_COMMIT)) {
        revokeRecoveryCodes(activation.getActivationId());
    }
    activationHistoryServiceBehavior.saveActivationAndLogChange(activation, externalUserId);
    callbackUrlBehavior.notifyCallbackListenersOnActivationChange(activation);
    RemoveActivationResponse response = new RemoveActivationResponse();
    response.setActivationId(activation.getActivationId());
    response.setRemoved(true);
    return response;
}
```

Reply:
But for this change, I would just use the original method, only to provide the correct logging message to the log ("Deactivating pending activation, ...").

Reply:
Ok, I will look into this.

Reply:
I cleaned up the code and improve code reuse.